sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -108,26 +108,30 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
   {
     std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
 
-    Command *NewCmd = static_cast<Command *>(NewEvent->getCommand());
+    // enqueueCommand() func below may throw an exception, so use unique_ptr
+    // here to avoid memory leak
+    std::unique_ptr<Command> NewCmd(
+        static_cast<Command *>(NewEvent->getCommand()));
+
     if (NewCmd) {
       // TODO: Check if lazy mode.
       EnqueueResultT Res;
-      bool Enqueued = GraphProcessor::enqueueCommand(NewCmd, Res);
+      bool Enqueued = GraphProcessor::enqueueCommand(NewCmd.get(), Res);
       if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
         throw runtime_error("Enqueue process failed.", PI_INVALID_OPERATION);
 
       // If there are no memory dependencies decouple and free the command.
       // Though, dismiss ownership of native kernel command group as it's
       // resources may be in use by backend and synchronization point here is
       // at native kernel execution finish.
-      if (NewCmd->MDeps.size() == 0 && NewCmd->MUsers.size() == 0) {
+      if (NewCmd.get()->MDeps.size() == 0 && NewCmd.get()->MUsers.size() == 0) {
         if (IsHostKernel)
-          static_cast<ExecCGCommand *>(NewCmd)->releaseCG();
+          static_cast<ExecCGCommand *>(NewCmd.get())->releaseCG();
 
         NewEvent->setCommand(nullptr);
-        delete NewCmd;
```

Comment:
Why is this line removed? Should it be replaced with `unique_ptr::reset()` to deallocate the command? I believe, it should as the command isn't tracked anymore.

Reply:
Fixed.